File: sdk/typescript/src/offline.ts
/**
 * NotaryOS Offline Verification - Ed25519 signature verification via JWKS.
 *
 * Uses Web Crypto API — still zero dependencies.
 *
 * Usage:
 *   import { OfflineVerifier } from 'notaryos/offline';
 *   const verifier = await OfflineVerifier.fromJWKS();
 *   const result = await verifier.verify(receipt);
 *   console.log(result.valid); // true
 *
 * @packageDocumentation
 */

export interface OfflineVerificationResult {
  valid: boolean;
  signatureOk: boolean;
  structureOk: boolean;
  reason: string;
  keyId: string;
}

interface JWK {
  kty: string;
  crv: string;
  alg: string;
  use: string;
  kid: string;
  x: string;
  status?: string;
}

/**
 * Offline verifier that fetches JWKS keys and verifies Ed25519 signatures locally.
 *
 * @example
 * ```typescript
 * const verifier = await OfflineVerifier.fromJWKS();
 * const result = await verifier.verify(receipt);
 * if (result.valid) {
 *   console.log('Verified offline!');
 * }
 * ```
 */
export class OfflineVerifier {
  private keys: Map<string, CryptoKey>;

  private constructor(keys: Map<string, CryptoKey>) {
    this.keys = keys;
  }

  /**
   * Create an OfflineVerifier by fetching JWKS from the server.
   *
   * @param baseUrl - API base URL (default: production)
   * @returns OfflineVerifier with cached public keys
   */
  static async fromJWKS(
    baseUrl = 'https://api.agenttownsquare.com'
  ): Promise<OfflineVerifier> {
    const url = `${baseUrl.replace(/\/+$/, '')}/.well-known/jwks.json`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch JWKS: ${response.statusText}`);
    }

    const jwks: { keys: JWK[] } = await response.json();
    const keys = new Map<string, CryptoKey>();

    for (const jwk of jwks.keys) {
      if (jwk.kty !== 'OKP' || jwk.crv !== 'Ed25519') continue;
      if (!jwk.kid || !jwk.x) continue;

      try {
        // Import the Ed25519 public key from JWK format
        const cryptoKey = await crypto.subtle.importKey(
          'jwk',
          { kty: 'OKP', crv: 'Ed25519', x: jwk.x },
          { name: 'Ed25519' },
          true,
          ['verify']
        );
        keys.set(jwk.kid, cryptoKey);
      } catch {
        // Skip keys that can't be imported (older runtimes)
      }
    }

    if (keys.size === 0) {
      throw new Error('No Ed25519 keys found in JWKS response');
    }

    return new OfflineVerifier(keys);
  }

  /**
   * Verify a receipt's signature offline using cached keys.
   *
   * @param receipt - Receipt object or dict
   * @returns OfflineVerificationResult
   */
  async verify(receipt: Record<string, unknown>): Promise<OfflineVerificationResult> {
    // Check structure
    const required = [
      'receipt_id', 'timestamp', 'agent_id', 'action_type',
      'payload_hash', 'signature', 'signature_type',
    ];
    const missing = required.filter((f) => !receipt[f]);
    if (missing.length > 0) {
      return {
        valid: false,
        signatureOk: false,
        structureOk: false,
        reason: `Missing required fields: ${missing.join(', ')}`,
        keyId: '',
      };
    }

    // Find the key
    const kid = (receipt.kid || receipt.key_id || '') as string;
    let cryptoKey = this.keys.get(kid);

    // Try prefix match
    if (!cryptoKey) {
      for (const [storedKid, storedKey] of this.keys) {
        if (storedKid.startsWith(kid.slice(0, 8)) || kid.startsWith(storedKid.slice(0, 8))) {
          cryptoKey = storedKey;
          break;
        }
      }
    }

    if (!cryptoKey) {
      return {
        valid: false,
        signatureOk: false,
        structureOk: true,
        reason: `Unknown key ID: ${kid}`,
        keyId: kid,
      };
    }

    // Reconstruct canonical message
    const canonical = buildCanonical(receipt);

    // Verify the signature
    try {
      const sigBytes = base64Decode(receipt.signature as string);
      const messageBytes = new TextEncoder().encode(canonical);

      const valid = await crypto.subtle.verify(
        'Ed25519',
        cryptoKey,
        sigBytes,
        messageBytes
      );

      return {
        valid,
        signatureOk: valid,
        structureOk: true,
        reason: valid ? 'Signature verified locally' : 'Signature mismatch',
        keyId: kid,
      };
    } catch (err) {
      return {
        valid: false,
        signatureOk: false,
        structureOk: true,
        reason: `Signature verification failed: ${(err as Error).message}`,
        keyId: kid,
      };
    }
  }

  /** Return all cached key IDs. */
  get keyIds(): string[] {
    return Array.from(this.keys.keys());
  }
}

// =============================================================================
// Internal helpers
// =============================================================================

function buildCanonical(receipt: Record<string, unknown>): string {
  const parts = [
    receipt.receipt_id || '',
    receipt.timestamp || '',
    receipt.agent_id || '',
    'notary',
    receipt.action_type || '',
    receipt.payload_hash || '',
    (receipt.previous_receipt_hash as string) || 'GENESIS',
  ];
  return parts.join('|');
}

function base64Decode(str: string): Uint8Array {
  // Handle both standard and URL-safe base64
  const b64 = str.replace(/-/g, '+').replace(/_/g, '/');
  const padding = '='.repeat((4 - (b64.length % 4)) % 4);
  const binary = atob(b64 + padding);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

File: sdk/go/notary/offline.go
package notary

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// OfflineVerificationResult holds the result of offline verification.
type OfflineVerificationResult struct {
	Valid       bool   `json:"valid"`
	SignatureOK bool   `json:"signature_ok"`
	StructureOK bool   `json:"structure_ok"`
	Reason      string `json:"reason"`
	KeyID       string `json:"key_id"`
}

// OfflineVerifier verifies Notary receipt signatures using cached Ed25519 public keys.
// It fetches keys from JWKS and performs all verification locally.
type OfflineVerifier struct {
	keys map[string]ed25519.PublicKey // kid -> 32-byte public key
}

// JWK represents a JSON Web Key (Ed25519 OKP).
type jwkKey struct {
	Kty    string `json:"kty"`
	Crv    string `json:"crv"`
	Alg    string `json:"alg"`
	Use    string `json:"use"`
	Kid    string `json:"kid"`
	X      string `json:"x"`
	Status string `json:"status,omitempty"`
}

type jwksResponse struct {
	Keys []jwkKey `json:"keys"`
}

// NewOfflineVerifier creates an OfflineVerifier by fetching JWKS from the server.
//
//	verifier, err := notary.NewOfflineVerifier("")
//	result := verifier.Verify(receiptMap)
//	fmt.Println(result.Valid)
func NewOfflineVerifier(baseURL string) (*OfflineVerifier, error) {
	if baseURL == "" {
		baseURL = DefaultBaseURL
	}

	url := strings.TrimRight(baseURL, "/") + "/.well-known/jwks.json"

	resp, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("JWKS fetch failed with status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read JWKS response: %w", err)
	}

	var jwks jwksResponse
	if err := json.Unmarshal(body, &jwks); err != nil {
		return nil, fmt.Errorf("failed to parse JWKS: %w", err)
	}

	keys := make(map[string]ed25519.PublicKey)
	for _, jwk := range jwks.Keys {
		if jwk.Kty != "OKP" || jwk.Crv != "Ed25519" {
			continue
		}
		if jwk.Kid == "" || jwk.X == "" {
			continue
		}

		// Decode base64url-encoded public key
		rawKey, err := base64.RawURLEncoding.DecodeString(jwk.X)
		if err != nil {
			// Try with padding
			rawKey, err = base64.URLEncoding.DecodeString(jwk.X)
			if err != nil {
				continue
			}
		}

		if len(rawKey) == ed25519.PublicKeySize {
			keys[jwk.Kid] = ed25519.PublicKey(rawKey)
		}
	}

	if len(keys) == 0 {
		return nil, fmt.Errorf("no Ed25519 keys found in JWKS response")
	}

	return &OfflineVerifier{keys: keys}, nil
}

// Verify checks a receipt's signature offline using cached Ed25519 keys.
func (v *OfflineVerifier) Verify(receipt map[string]any) *OfflineVerificationResult {
	// Check structure
	required := []string{
		"receipt_id", "timestamp", "agent_id", "action_type",
		"payload_hash", "signature", "signature_type",
	}
	var missing []string
	for _, field := range required {
		val, ok := receipt[field]
		if !ok || val == nil || val == "" {
			missing = append(missing, field)
		}
	}
	if len(missing) > 0 {
		return &OfflineVerificationResult{
			Valid:       false,
			SignatureOK: false,
			StructureOK: false,
			Reason:      fmt.Sprintf("Missing required fields: %s", strings.Join(missing, ", ")),
		}
	}

	// Find the key
	kid := getString(receipt, "kid")
	if kid == "" {
		kid = getString(receipt, "key_id")
	}

	pubKey, found := v.keys[kid]
	if !found {
		// Try prefix match
		for storedKid, storedKey := range v.keys {
			if len(kid) >= 8 && len(storedKid) >= 8 {
				if strings.HasPrefix(storedKid, kid[:8]) || strings.HasPrefix(kid, storedKid[:8]) {
					pubKey = storedKey
					kid = storedKid
					found = true
					break
				}
			}
		}
	}

	if !found {
		return &OfflineVerificationResult{
			Valid:       false,
			SignatureOK: false,
			StructureOK: true,
			Reason:      fmt.Sprintf("Unknown key ID: %s", kid),
			KeyID:       kid,
		}
	}

	// Reconstruct canonical message
	canonical := buildCanonical(receipt)

	// Decode signature
	sigStr := getString(receipt, "signature")
	sigBytes, err := base64.StdEncoding.DecodeString(sigStr)
	if err != nil {
		// Try URL-safe base64
		sigBytes, err = base64.RawURLEncoding.DecodeString(sigStr)
		if err != nil {
			return &OfflineVerificationResult{
				Valid:       false,
				SignatureOK: false,
				StructureOK: true,
				Reason:      fmt.Sprintf("Failed to decode signature: %v", err),
				KeyID:       kid,
			}
		}
	}

	// Verify Ed25519 signature
	valid := ed25519.Verify(pubKey, []byte(canonical), sigBytes)

	reason := "Signature verified locally"
	if !valid {
		reason = "Signature mismatch"
	}

	return &OfflineVerificationResult{
		Valid:       valid,
		SignatureOK: valid,
		StructureOK: true,
		Reason:      reason,
		KeyID:       kid,
	}
}

// KeyIDs returns all cached key IDs.
func (v *OfflineVerifier) KeyIDs() []string {
	ids := make([]string, 0, len(v.keys))
	for kid := range v.keys {
		ids = append(ids, kid)
	}
	return ids
}

func buildCanonical(receipt map[string]any) string {
	prevHash := getString(receipt, "previous_receipt_hash")
	if prevHash == "" {
		prevHash = "GENESIS"
	}

	parts := []string{
		getString(receipt, "receipt_id"),
		getString(receipt, "timestamp"),
		getString(receipt, "agent_id"),
		"notary",
		getString(receipt, "action_type"),
		getString(receipt, "payload_hash"),
		prevHash,
	}
	return strings.Join(parts, "|")
}

func getString(m map[string]any, key string) string {
	v, ok := m[key]
	if !ok || v == nil {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return fmt.Sprintf("%v", v)
	}
	return s
}
